### PR TITLE
Seq.toArray via custom ArrayBuilder

### PR DESCRIPTION
**Pull request by T-Gro**
Seq.toArray, besides its direct use, also serves as workhorse for other use cases:
- Seq sorting
- All the *back functions in Seq
- Non-trivial array `[||]` comprehensions get compiled into building a sequence and then calling Seq.toArray ( https://github.com/dotnet/fsharp/blob/main/src/Compiler/Checking/CheckComputationExpressions.fs#L2252 ) 

In the fallback scenario (seq is neither list, nor ICollection, nor Array), the following code was used to create and array out of a sequence:
```
let res = ResizeArray<_>(source)
res.ToArray()
```

For this specialized scenario of building an array, using ResizeArray has the 2 disadvantages (coming from the fact that it is a general collection for a variety of use cases, not because it is badly implemented):
- Is defensive and never returns the underlying array, even when it exactly matches current count
- Always means additional heap allocation for the ResizeArray itself

Benchmark results are available at https://github.com/T-Gro/SeqToArrayBenchmark , showing improvements in time and memory for smaller sizes, especially powers of two (due to point no. 2). 
https://github.com/T-Gro/SeqToArrayBenchmark/blob/main/BenchmarkResults.md

TLDR: The chosen `ManualSpecialCase0` method can be: 10-20% faster for smaller sizes, and 3% faster for bigger size.

I also had other assumptions, but they did not outperform the simpler one in place right now:
- ResizeArray keeps throwing away the smaller arrays it previously allocated. It needs to do that to provide O(1) lookup, but ArrayBuilder does not need that. However, having an F# list of increasing arrays (the up to date one always at the head) did not perform better for me.

- Initial building until a small size (e.g. for the first 8 items) could be done fully in stack using stackalloc.
This was not proven to be faster than then chosen `ManualSpecialCase0`, but it does avoid allocating compared to building the array dynamically.

### CODE REVIEW

**Comment by vzarytovskii** [comment]
Could you plese include benchmark results in comment or OP?

**Review by abonie — APPROVED** [review]

File: src/FSharp.Core/seq.fs [hunk 1]
```diff
@@ -18,6 +18,35 @@ open Microsoft.FSharp.Primitives.Basics
...
+    [<NoEquality>]
+    [<NoComparison>]
+    type ArrayBuilder<'T> =
+        {
+            mutable currentCount: int
+            mutable currentArray: 'T array
+        }
+
+    let inline addToBuilder (item: 'T) (builder: byref<ArrayBuilder<'T>>) =
+        match builder.currentCount = builder.currentArray.Length with
```
Comment (0101) [inline src/FSharp.Core/seq.fs:34]: Any reason for `match .. true | false` instead of `if else`?

**Review by 0101 — APPROVED** [review]

File: src/FSharp.Core/seq.fs (hunk 1 above)
Comment (T-Gro) [reply src/FSharp.Core/seq.fs:34]: Feels more condensed this way, but I am agnostic.

**Review by KevinRansom — APPROVED** [review]
